### PR TITLE
Contract compilation debugging infrastructure

### PR DESCRIPTION
**Pull request by olonho**

### CODE REVIEW

**Comment by nagisa** [comment]
My overall take is that we should approach this from a different angle. I think we should look into making it possible to dump the serialized `UniversalArtifact` from a) RocksDB of nearcore, via state-viewer perhaps; or b) generate one from `.wasm` on the fly with `near-vm-runner-standalone` or similar.

This would then allow a whole swath of tools to be constructed that can deserialize the `UniversalArtifact` and dump all sorts of interesting information from it (a-la `dumpelf(1)`). We could also investigate a tool to convert these artifacts to elf so that the tooling that exists for ELFs can be reused directly (shouldn't be too hard to achieve via e.g. the `object` crate)

**Comment by olonho** [comment]
We can dump serialized artefact, and it makes lot of sense to make changes you suggest, but it's pretty hard to disassemble it to do in depth code quality analysis.

**Comment by olonho** [comment]
And to implement conversion tools we need API to access function length from artefacts + provide ability to store symbolic information.

File: lib/engine-staticlib/src/artifact.rs [hunk 1]
```diff
@@ -477,6 +477,10 @@ impl Artifact for StaticlibArtifact {
         &self.finished_functions
     }
 
+    fn finished_functions_lengths(&self) -> &BoxedSlice<LocalFunctionIndex, usize> {
+        todo!("unimpl");
```
Comment (matklad) [inline lib/engine-staticlib/src/artifact.rs:481]:
```suggestion
        unimplemented!("");
```

**Comment by matklad** [comment]
Yeah, it does seem that we obviously want both:

* high level tool which takes artifact and dumps it into some format interoperable with other tools, or just directly inspects it
* low level API which tells where the functions are 

Though, I'd probably avoid committing the code referring to specific files on Nikolay's hard-drive :-)

**Comment by nagisa** [comment]
I think one of the important parts of my message became implicit. One of the adjustments I am suggesting is that we want to deal with `UniversalArtifact`s only. That is, the concrete implementation of the `Artifact` trait and not the trait object. I don't anticipate us implementing anything similar for the staticlib or dylib engines either way – and the function length information is readily available as part of the `UniversalArtifact` as a field.

So in practice doing something along the lines of `UniversalArtifact::deserialize(engine, bytes).finished_function_lengths()` would be a reasonably nice approach to obtain the information about functions in the disassembled code without modifying the `Artifact` trait.

**Comment by olonho** [comment]
Made it draft for now, as https://github.com/near/wasmer/pull/98 is better.

**Comment by nagisa** [comment]
Since we landed the alternative implementation I'm going to close this for the time being.

**Pull request closed without merging**